### ISSUE TITLE
Crash if we reset QQmlApplicationEngine

### ISSUE THREAD

**Issue by misery** [KDAB/GammaRay#152]
For debugging purpose we have a shortcut to reload the entire QQmlApplicationEngine. This works very well until we attach gammaray. If we reload the engine gammaray will crash.

Our init/reload code:
```
QScopedPointer<QQmlApplicationEngine> mEngine;

void App::init()
{
	mEngine.reset(new QQmlApplicationEngine());

#ifndef QT_NO_DEBUG
	mEngine->rootContext()->setContextProperty("plugin", this);
#endif

	auto providerSettings = &AppSettings::getInstance().getProviderSettings();
	auto historySettings = &AppSettings::getInstance().getHistorySettings();

	auto providerModel = new ProviderModel(providerSettings, mEngine.data());
	mEngine->rootContext()->setContextProperty("providerModel", providerModel);

	auto historyModel = new HistoryModel(historySettings, providerSettings, mEngine.data());
	mEngine->rootContext()->setContextProperty("historyModel", historyModel);

	mEngine->load(getMainfile());
}
```

Qt 5.5.0
GammaRay: master (34e4d4784505067136162bb09bccc3b4d1d78146)

```
==31992==ERROR: AddressSanitizer: SEGV on unknown address 0x7f7b3d00020c (pc 0x7f7ba89d151f bp 0x7f7ba705e780 sp 0x7f7ba705e750 T7)
    #0 0x7f7ba89d151e in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:234
    #1 0x7f7ba89d1432 in GammaRay::TranslatorWrapper::translate(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:217
    #2 0x7f7bc7e3a800 in QCoreApplication::translate(char const*, char const*, char const*, int) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:1953
    #3 0x7f7bc9756196 in QQmlImportDatabase::tr(char const*, char const*, int) [...]/qt/qtdeclarative/src/qml/qml/qqmlimport_p.h:141
    #4 0x7f7bc974e639 in QQmlImportNamespace::resolveType(QQmlTypeLoader*, QHashedStringRef const&, int*, int*, QQmlType**, QString*, QList<QQmlError>*) [...]/qt/qtdeclarative/src/qml/qml/qqmlimport.cpp:803
    #5 0x7f7bc974db75 in QQmlImportsPrivate::resolveType(QHashedStringRef const&, int*, int*, QQmlType**, QList<QQmlError>*) [...]/qt/qtdeclarative/src/qml/qml/qqmlimport.cpp:724
    #6 0x7f7bc974c263 in QQmlImports::resolveType(QHashedStringRef const&, QQmlType**, int*, int*, QQmlImportNamespace**, QList<QQmlError>*) const [...]/qt/qtdeclarative/src/qml/qml/qqmlimport.cpp:525
    #7 0x7f7bc971265a in QQmlTypeData::resolveTypes() [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:2314
    #8 0x7f7bc9711ca0 in QQmlTypeData::allDependenciesDone() [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:2217
    #9 0x7f7bc970c288 in QQmlTypeLoader::setData(QQmlDataBlob*, QQmlDataBlob::Data const&) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1202
    #10 0x7f7bc970c1b9 in QQmlTypeLoader::setData(QQmlDataBlob*, QQmlFile*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1191
    #11 0x7f7bc970b962 in QQmlTypeLoader::loadThread(QQmlDataBlob*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1069
    #12 0x7f7bc970b055 in QQmlTypeLoader::load(QQmlDataBlob*, QQmlTypeLoader::Mode) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:927
    #13 0x7f7bc970e384 in QQmlTypeLoader::getType(QUrl const&, QQmlTypeLoader::Mode) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1613
    #14 0x7f7bc9712b26 in QQmlTypeData::resolveTypes() [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:2355
    #15 0x7f7bc9711ca0 in QQmlTypeData::allDependenciesDone() [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:2217
    #16 0x7f7bc970c288 in QQmlTypeLoader::setData(QQmlDataBlob*, QQmlDataBlob::Data const&) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1202
    #17 0x7f7bc970c1b9 in QQmlTypeLoader::setData(QQmlDataBlob*, QQmlFile*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1191
    #18 0x7f7bc970b962 in QQmlTypeLoader::loadThread(QQmlDataBlob*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1069
    #19 0x7f7bc970b055 in QQmlTypeLoader::load(QQmlDataBlob*, QQmlTypeLoader::Mode) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:927
    #20 0x7f7bc970e384 in QQmlTypeLoader::getType(QUrl const&, QQmlTypeLoader::Mode) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1613
    #21 0x7f7bc9712b26 in QQmlTypeData::resolveTypes() [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:2355
    #22 0x7f7bc9711ca0 in QQmlTypeData::allDependenciesDone() [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:2217
    #23 0x7f7bc970c288 in QQmlTypeLoader::setData(QQmlDataBlob*, QQmlDataBlob::Data const&) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1202
    #24 0x7f7bc970c1b9 in QQmlTypeLoader::setData(QQmlDataBlob*, QQmlFile*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1191
    #25 0x7f7bc970b962 in QQmlTypeLoader::loadThread(QQmlDataBlob*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1069
    #26 0x7f7bc970ad18 in QQmlTypeLoaderThread::loadThread(QQmlDataBlob*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:813
    #27 0x7f7bc9715d90 in void QQmlThread::callMethodInThread<QQmlDataBlob*, QQmlDataBlob*, QQmlTypeLoaderThread>(void (QQmlTypeLoaderThread::*)(QQmlDataBlob*), QQmlDataBlob* const&)::I::call(QQmlThread*) [...]/qt/qtdeclarative/src/qml/../../include/QtQml/5.5.0/QtQml/private/../../../../../src/qml/qml/ftw/qqmlthread_p.h:154
    #28 0x7f7bc9785b82 in QQmlThreadPrivate::threadEvent() [...]/qt/qtdeclarative/src/qml/qml/ftw/qqmlthread.cpp:190
    #29 0x7f7bc9785903 in QQmlThreadPrivate::event(QEvent*) [...]/qt/qtdeclarative/src/qml/qml/ftw/qqmlthread.cpp:128
    #30 0x7f7bca1a5663 in QApplicationPrivate::notify_helper(QObject*, QEvent*) [...]/qt/qtbase/src/widgets/kernel/qapplication.cpp:3717
    #31 0x7f7bca1a2dc2 in QApplication::notify(QObject*, QEvent*) [...]/qt/qtbase/src/widgets/kernel/qapplication.cpp:3161
    #32 0x7f7bc7e386b1 in QCoreApplication::notifyInternal(QObject*, QEvent*) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:965
    #33 0x7f7bc7e3c3be in QCoreApplication::sendEvent(QObject*, QEvent*) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.h:224
    #34 0x7f7bc7e39a2d in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:1593
    #35 0x7f7bc7e393ad in QCoreApplication::sendPostedEvents(QObject*, int) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:1451
    #36 0x7f7bc7eadfeb in postEventSourceDispatch(_GSource*, int (*)(void*), void*) [...]/qt/qtbase/src/corelib/kernel/qeventdispatcher_glib.cpp:271
    #37 0x7f7bc5b4cdc6 in g_main_context_dispatch (/usr/lib/libglib-2.0.so.0+0x49dc6)
    #38 0x7f7bc5b4d01f  (/usr/lib/libglib-2.0.so.0+0x4a01f)
    #39 0x7f7bc5b4d0cb in g_main_context_iteration (/usr/lib/libglib-2.0.so.0+0x4a0cb)
    #40 0x7f7bc7eae7d7 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) [...]/qt/qtbase/src/corelib/kernel/qeventdispatcher_glib.cpp:420
    #41 0x7f7bc7e3526f in QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) [...]/qt/qtbase/src/corelib/kernel/qeventloop.cpp:128
    #42 0x7f7bc7e35563 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) [...]/qt/qtbase/src/corelib/kernel/qeventloop.cpp:204
    #43 0x7f7bc7bf574f in QThread::exec() [...]/qt/qtbase/src/corelib/thread/qthread.cpp:503
    #44 0x7f7bc9785974 in QQmlThreadPrivate::run() [...]/qt/qtdeclarative/src/qml/qml/ftw/qqmlthread.cpp:141
    #45 0x7f7bc7bfcdf2 in QThreadPrivate::start(void*) [...]/qt/qtbase/src/corelib/thread/qthread_unix.cpp:331
    #46 0x7f7bc706b4a3 in start_thread (/usr/lib/libpthread.so.0+0x74a3)
    #47 0x7f7bc678713c in __clone (/usr/lib/libc.so.6+0xe913c)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:234 in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const
Thread T7 (QQmlThread) created by T0 here:
    #0 0x430d29 in __interceptor_pthread_create ([...]/build/src/App+0x430d29)
    #1 0x7f7bc7bfd662 in QThread::start(QThread::Priority) [...]/qt/qtbase/src/corelib/thread/qthread_unix.cpp:609
    #2 0x7f7bc9785ce0 in QQmlThread::startup() [...]/qt/qtdeclarative/src/qml/qml/ftw/qqmlthread.cpp:220
    #3 0x7f7bc970a7d7 in QQmlTypeLoaderThread::QQmlTypeLoaderThread(QQmlTypeLoader*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:728
    #4 0x7f7bc970e0b5 in QQmlTypeLoader::QQmlTypeLoader(QQmlEngine*) [...]/qt/qtdeclarative/src/qml/qml/qqmltypeloader.cpp:1570
    #5 0x7f7bc96ccbc7 in QQmlEnginePrivate::QQmlEnginePrivate(QQmlEngine*) [...]/qt/qtdeclarative/src/qml/qml/qqmlengine.cpp:592
    #6 0x7f7bc976f232 in QQmlApplicationEnginePrivate::QQmlApplicationEnginePrivate(QQmlEngine*) [...]/qt/qtdeclarative/src/qml/qml/qqmlapplicationengine.cpp:44
    #7 0x7f7bc976fd4a in QQmlApplicationEngine::QQmlApplicationEngine(QObject*) [...]/qt/qtdeclarative/src/qml/qml/qqmlapplicationengine.cpp:196
    #8 0x7f7bcb12b2a0 in UIPlugInQml::init() [...]/Source/src/qml/UIPlugInQml.cpp:44:20
    #9 0x7f7bcb130ab2 in UIPlugInQml::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) [...]/build/src/qml/moc_UIPlugInQml.cpp:78:17
    #10 0x7f7bc7e6f57c in QMetaCallEvent::placeMetaCall(QObject*) [...]/qt/qtbase/src/corelib/kernel/qobject.cpp:486
    #11 0x7f7bc7e70633 in QObject::event(QEvent*) [...]/qt/qtbase/src/corelib/kernel/qobject.cpp:1246
    #12 0x7f7bca1a5663 in QApplicationPrivate::notify_helper(QObject*, QEvent*) [...]/qt/qtbase/src/widgets/kernel/qapplication.cpp:3717
    #13 0x7f7bca1a2dc2 in QApplication::notify(QObject*, QEvent*) [...]/qt/qtbase/src/widgets/kernel/qapplication.cpp:3161
    #14 0x7f7bc7e386b1 in QCoreApplication::notifyInternal(QObject*, QEvent*) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:965
    #15 0x7f7bc7e3c3be in QCoreApplication::sendEvent(QObject*, QEvent*) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.h:224
    #16 0x7f7bc7e39a2d in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:1593
    #17 0x7f7bc7e393ad in QCoreApplication::sendPostedEvents(QObject*, int) [...]/qt/qtbase/src/corelib/kernel/qcoreapplication.cpp:1451
    #18 0x7f7bc7eadfeb in postEventSourceDispatch(_GSource*, int (*)(void*), void*) [...]/qt/qtbase/src/corelib/kernel/qeventdispatcher_glib.cpp:271
    #19 0x7f7bc5b4cdc6 in g_main_context_dispatch (/usr/lib/libglib-2.0.so.0+0x49dc6)

==31992==ABORTING
```

**Comment by krf** [comment]
@misery: Could you try the following patch and report back if it fixes the issue for you?

```diff
diff --git a/plugins/translatorinspector/translatorwrapper.cpp b/plugins/translatorinspector/translatorwrapper.cpp
index 3728d9d..685956e 100644
--- a/plugins/translatorinspector/translatorwrapper.cpp
+++ b/plugins/translatorinspector/translatorwrapper.cpp
@@ -235,7 +235,7 @@ QString TranslatorWrapper::translateInternal(const char *context,
 }
 const QTranslator *TranslatorWrapper::translator() const
 {
-  return m_wrapped == 0 ? this : m_wrapped;
+  return m_wrapped ? this : m_wrapped;
 }
 
 FallbackTranslator::FallbackTranslator(QObject *parent)
diff --git a/plugins/translatorinspector/translatorwrapper.h b/plugins/translatorinspector/translatorwrapper.h
index 294cc41..755945e 100644
--- a/plugins/translatorinspector/translatorwrapper.h
+++ b/plugins/translatorinspector/translatorwrapper.h
@@ -30,6 +30,7 @@
 #define TRANSLATORWRAPPER_H
 
 #include <QAbstractItemModel>
+#include <QPointer>
 #include <QTranslator>
 
 namespace GammaRay {
@@ -103,7 +104,7 @@ class TranslatorWrapper : public QTranslator
     const QTranslator *translator() const;
 
   private:
-    QTranslator *m_wrapped;
+    QPointer<QTranslator> m_wrapped;
     TranslationsModel *m_model;
 
     QString translateInternal(const char *context, const char *sourceText
```

**Comment by misery** [comment]
```
==879==ERROR: AddressSanitizer: stack-overflow on address 0x7ffd717c1ff8 (pc 0x7ff47306a59f bp 0x7ffd717c2030 sp 0x7ffd717c2000 T0)
    #0 0x7ff47306a59e in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:231
    #1 0x7ff47306a3e1 in GammaRay::TranslatorWrapper::translate(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:214
    #2 0x7ff47306a5cd in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:231
    #3 0x7ff47306a3e1 in GammaRay::TranslatorWrapper::translate(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:214
    #4 0x7ff47306a5cd in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:231
    #5 0x7ff47306a3e1 in GammaRay::TranslatorWrapper::translate(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:214
    #6 0x7ff47306a5cd in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:231
    #7 0x7ff47306a3e1 in GammaRay::TranslatorWrapper::translate(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:214
    #8 0x7ff47306a5cd in GammaRay::TranslatorWrapper::translateInternal(char const*, char const*, char const*, int) const /[...]/GammaRay/plugins/translatorinspector/translatorwrapper.cpp:231

[...]
```

**Comment by krf** [comment]
Hm. Makes sense. Not a proper fix.

I don't know this code enough / don't have time to work on it. Maybe you want to dive into the code base and try to make sense of it?

**Comment by vkrause** [comment]
Should be fixed in 2.4 and master.
